feat(ai): add safeValidateUIMessages

fveiraswww opened the pull request:
## Background

The current validateUIMessages throws on validation failure. For API endpoints and similar contexts it's ergonomically preferable to use a safe-return pattern (like zod's safeParse) to avoid try/catch and get a typed success/failure result. 

see context in #8673 

## Summary

- Add `safeValidateUIMessages`
- Returns a discriminated union: `{ success: true, data }` || `{ success: false, reason, error }`

## Manual Verification

- Ran unit tests in packages/ai

## Tasks

- [x] Tests have been added / updated (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)
- [x] Formatting issues have been fixed (run `pnpm prettier-fix` in the project root)
- [x] I have reviewed this pull request (self-review)

## Related Issues

fixes #8673

fveiraswww committed:
feat(ai): add safeValidateUIMessages

gr2m reviewed (commented):
could you add documentation as we have for `validateUIMessages()`: https://ai-sdk.dev/docs/reference/ai-sdk-core/validate-ui-messages

gr2m commented:
thanks for the pull request 👍🏼 if you could comment on the issue that you are working on, that would be great! For one, it allows us to assign the issue to you, and second people know that somebody is already working on it and won't duplicate your work

lgrammel commented on packages/ai/src/ui/validate-ui-messages.ts:
```diff
@@ -286,3 +286,61 @@ export async function validateUIMessages<UI_MESSAGE extends UIMessage>({
+      error: err,
+    };
+  }
+}
```
instead of try catch around validateUIMessages, can we look into using safeValidate as default and have validateUIMessages wrap safeValidate so we do not throw/catch errors unecessarily

fveiraswww replied:
sounds good, so something like:

```ts
export async function validateUIMessages<>(...): Promise<Array<UI_MESSAGE>> {  
  const res = await safeValidateUIMessages<UI_MESSAGE>(args);  
  if (!res.success) throw res.error;  
  return res.data;  
}
```

that way `safeValidateUIMessages` is the default and `validateUIMessages` just wraps it

fveiraswww commented on packages/ai/src/ui/validate-ui-messages.test.ts:
```diff
@@ -1106,3 +1109,139 @@ describe('validateUIMessages', () => {
+    });
+
+    expect(result.success).toBe(true);
+    if (result.success) {
```
if I remove the if I get a ts error because expect().toBe(true) doesn't narrow the type. I may be missing something.

`Property 'data' does not exist on type '{ success: false; error: Error; }'`

the workarounds I see look a bit ugly, what do you suggest?

fveiraswww committed:
refactor(ai): make validateUIMessages wrap safeValidateUIMessages

fveiraswww commented:
@gr2m I added the docs following the same style as validateUIMessages and applied the other fixes

@lgrammel I refactored validateUIMessages to wrap safeValidateUIMessages as suggested ✅

gr2m committed:
Merge branch 'main' into feat/add-safe-validate-ui-messages

fveiraswww committed:
refactor(test): add expectToBe assertion

gr2m committed:
Merge branch 'main' into feat/add-safe-validate-ui-messages
